Modify the logic for deleting variations

usbcisthebest opened the pull request:
## Summary 

Implements the following expected behavior which fixes an issue with Subscriptions:
1. Switch from variable subscription to non-variable one will delete variations.
2. Switch from variable subscription to variable product and vice-versa will NOT delete any variations.

This shall have no side-effects in WooCommerce Core.

### All Submissions:

* [x] Have you followed the [WooCommerce Contributing guideline](https://github.com/woocommerce/woocommerce/blob/master/.github/CONTRIBUTING.md)?
* [x] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/)?
* [x] Have you checked to ensure there aren't other open [Pull Requests](../../pulls) for the same update/change?

<!-- Mark completed items with an [x] -->

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

<!-- Describe the changes made to this Pull Request and the reason for such changes. -->

Closes https://github.com/woocommerce/woocommerce/issues/26959.
Closes https://github.com/woocommerce/woocommerce-subscriptions/issues/3672

### How to test the changes in this Pull Request:

This was originally pushed to WC Subscriptions but it makes more sense to add it directly to WooCommerce. Testing instructions are therefore available in WC Subscriptions PR: https://github.com/woocommerce/woocommerce-subscriptions/pull/3795.

### Other information:

* [x] Have you added an explanation of what your changes do and why you'd like us to include them?
* [ ] Have you written new tests for your changes, as applicable?
* [ ] Have you successfully run tests with your changes locally?

<!-- Mark completed items with an [x] -->

### Changelog entry

> Improves the logic for deleting variations when a product type is changed

Commit:
Modify the logic for deleting variations

Implements the following expected behaviour:
1. Switch from variable subscription to non-variable one will delete variations.
2. Switch from variable subscription to variable product and vice-versa will NOT delete any variations.

This shall have no side-effects in WooCommerce Core.

rrennick reviewed (changes requested):
@codestor4 This has the potential to change behaviour for third party product types. This could be implemented as a filter that defaulted to the current logic. WC Subs could use that filter to bypass removing the variations.

Commit:
Introduce filters to filter $from and $to

This will introduce two new filters:

woocommerce_from_product_type_changed
woocommerce_to_product_type_changed

to filter the $from and $to variables respectively.

This will be useful in WCS pr_3732!

codestor4 commented:
Thanks for the feedback @rrennick. I have made the necessary changes now.

The newly introduced filters will be used as follows in https://github.com/woocommerce/woocommerce-subscriptions/pull/3732:

```
add_filter( 'woocommerce_from_product_type_changed', array( __CLASS__, 'maybe_keep_variations' ) );
add_filter( 'woocommerce_to_product_type_changed', array( __CLASS__, 'maybe_keep_variations' ) );

/**
 * Prevent variations from being deleted when product type changes from variable product to another variable product.
 *
 * @since 3.0.8
 *
 * @return string $product_type
 */
public static function maybe_keep_variations( $product_type ) {
	if ( strpos( $product_type, 'variable' ) ) {
		$product_type = 'variable';
	}
	return $product_type;
}
```

I've tested this and it works. I'll be adding it to the pull request mentioned above after this current pull request gets merged.

Commit:
Convert the two filters into a single filter

Removing the two newly introduced filters and replacing it with one single filter: woocommerce_delete_variations_on_product_type_change.

codestor4 commented:
Update: I have removed the two filters and replaced it with one. It still achieves the same end goal.

codestor4 commented:
Hey @rrennick, no rush but could you review this once again when you have some time? Thx :)

Konamiman commented:
Hi @codestor4, the change looks fine but it would be great if you could add a documentation comment for the newly added filter. See an example here: https://github.com/woocommerce/woocommerce/blob/master/includes/admin/meta-boxes/views/html-variation-admin.php#L165-L174 (please use version 4.9 for the `@since` field).

Commit:
Inline doc for the new filter

Updated @since to 4.9 and added an inline doc for the new filter
'woocommerce_delete_variations_on_product_type_change'

codestor4 commented:
@Konamiman I've made the requested changes. Please review and let me know! 👍

abhishek-pokhriyal commented on includes/class-wc-post-data.php:
```diff
@@ -121,13 +121,24 @@ public static function delete_product_query_transients() {
 	 * Handle type changes.
 	 *
-	 * @since 3.0.0
+	 * @since 4.9.0
```
I don't think this should be changed. If really needed, another since tag can be added to describe changes made to the method in this 4.9.0 (but I don't think that's needed because of the since tag on the filter).

codestor4 replied:
Oh! right. Sorry I didn't think it through 🤦‍♂️

abhishek-pokhriyal commented on includes/class-wc-post-data.php:
```diff
@@ -121,13 +121,24 @@ public static function delete_product_query_transients() {
+		 *
+		 * @param bool       A boolean value of true will delete the variations.
+		 * @param WC_Product $product Product data.
+		 * @return string    $from    Origin type.
```
"Origin" or "Original"?

codestor4 replied:
"Origin" is the intentional use. It's directly copied from the "Origin type" text [above](https://github.com/woocommerce/woocommerce/blame/34563c22bb302a969da0d966884c4d17b020a950/includes/class-wc-post-data.php#L127) (line 127). IMO "origin" and "new" makes sense in the context of "from" and "to. What do you think?

abhishek-pokhriyal replied:
Ah didn't notice the parameter comment in the method doc block. Keeping it consistent makes sense so `Origin type` is fine.

Commit:
Revert functon's @since version num back to 3.0

codestor4 commented:
@abhishek-pokhriyal It's ready for a review 👍

Konamiman commented:
Approved. @rrennick do you want to take a second look before merging?

rrennick commented:
@codestor4 Sorry for the delays in getting this through review.
